safe_gpu/safe_gpu.py

```diff
@@ -140,13 +150,52 @@ def allocate_gpus(self, gpu_device_numbers):
 
         time.sleep(self.debug_sleep)
 
+        if not self.placeholder_fn:
+            self.logger.info(f"No placeholder imposed (placeholder_fn = {self.placeholder_fn}).")
+
         try:
             self.placeholders = [self.placeholder_fn(device_no) for device_no in range(len(gpu_device_numbers))]
         except RuntimeError:
-            self.logger.error('Failed to acquire placeholder, truly marvellous. Race condition with someone not using `GPUOwner?`')
+            self.logger.error("""
+                              Failed to acquire placeholder.
+                              Race condition with someone not using `GPUOwner` ?
+                              Or CUDA was alrady initialized when calling `claim_gpus()`?
+                              """)
             raise
 
 
 def claim_gpus(nb_gpus=1, placeholder_fn=pytorch_placeholder, logger=None, debug_sleep=0.0):
+    """
+    Allocate the GPUs.
+
+    :param nb_gpus: number of GPUs to be allocated.
+    :param placeholder_fn: Placeholder function, it creates CUDA context at particular GPU.
+                           Typically by creating a ``dummy'' Tensor that gets stored till the end
+                           of the program life-time.
+    :param logger: argument for passing in the logging module or object (e.g. `logger=logging`)
+    :param debug_sleep: delay between setting `os.environ['CUDA_VISIBLE_DEVICES']` and executing `placeholder_fn` calls.
+
+    IMPORTANT NOTICE:
+
+    This function must be called before initializing CUDA:
+    - The `import torch` can be done before calling `claim_gpus()`.
+    - But `torch.cuda.is_available()` should not be called before `claim_gpus()`,
+      otherwise setting `CUDA_VISIBLE_DEVICES`in `GPUOwner::allocate_gpus()` will work well.
+
+    """
+
     global gpu_owner
     gpu_owner = GPUOwner(nb_gpus, placeholder_fn, logger, debug_sleep)
+
+def release_gpus():
+    global gpu_owner
+    gpu_owner.logger.info(f"RELEASING CUDA DEVICES (CUDA_VISIBLE_DEVICES={os.environ['CUDA_VISIBLE_DEVICES']}):")
+
+    if gpu_owner.placeholder_fn is pycuda_placeholder:
```

Comment:
1) Do we want to do the selection here? In principle, I would like it more have the release function registered in `GPUOwner`, so that this function stays closed for editing (if more release mechanisms are introduced).
2) Is it ok that nothing happens for other placeholders? I am not saying this should crash, but a Warning would seem appropriate. Combined with the above point 1), we could have a default "no-op" release function which just warns the user that it didn't really release anything.

Reply:
Okay, i'll hook gpu-release into `GPUOwner` destructor then...

For other placeholders, the placeholder variables can be released too,
but that would not release the GPU itself IMO...